### PR TITLE
Update WCSLIB to version 8.3

### LINKED ISSUE #16452: ANN: New WCSLIB 8.3 released
New WCSLIB release found.

Version: 8.3 (2024/05/14)

#### Change log

WCSLIB version 8.3 (2024/05/14)
-------------------------------

* C library

  - Until now, wcsset() always operated unconditionally - the wcsprm
    struct was set or reset regardless of its current state.  Likewise
    the various *set() functions for the other structs.

    However, in some situations, particularly in threaded applications,
    it is desirable to have wcsset(), etc., check the state of the
    struct and return immediately if it has already been set.  This may
    now be accomplished by setting wcsprm::flag == 1 (instead of 0)
    before calling wcsset().  Likewise for the other structs.  This sets
    a "bypass" flag within the struct itself.

  - A new function, wcsenq(), queries the state of the wcsprm struct,
    specifically whether WCSLIB is managing its memory, whether the
    struct has been set, and whether or not it is in bypass mode, and
    whether it is self-consistent.  There are corresponding functions
    for the other structs: celenq(), disenq(), linenq(), prjenq(),
    spcenq(), and tabenq().  Please refer to the WCSLIB manual.

  - In the C test suite, modified twcs to test wcsenq() and also
    wcsset() in bypass mode.

  - Quelled nuisance compiler warnings in wcsbth(), wcspih(), wcsp2s(),
    and wcshdo().

* Fortran wrappers

  - Interprocedural Link Time Optimization (LTO), when used with the
    strict compiler options required by some Linux distributions, may
    place more stringent requirements on mixing code written in
    different languages.  Specifically, as far as WCSLIB is concerned,
    this applies for Fortran calling C (or vice versa) where the
    function parameter list includes a character variable.

    It is important to note that the existing Fortran wrappers work as
    they did before, with or without LTO, that issues only arise when
    strict LTO compiler options are enabled, and that wrappers without
    a character argument are unaffected.

    Fortran 2003 introduced the "language-binding-spec" attribute using
    the keyword BIND.  The INTERFACE block for a procedure may be given
    the BIND(C) attribute to specify the interface of an external,
    interoperable C function.  Use of this BIND(C) attribute is now
    virtually mandated by LTO (with said strict compiler options).

    The WCSLIB Fortran wrappers are written in C, e.g. wcs_f.c, with a
    Fortran-compliant interface, and these C functions are intended to
    be called directly from Fortran applications.  Three options were
    considered for achieving strict LTO compliance:

      1) Require that all existing Fortran applications be modified to
         conform to the Fortran 2003 language binding specification via
         the addition of INTERFACE blocks bearing the BIND(C) attribute
         for the existing wrappers.

         This option is clearly untenable.

      2) Rewrite the WCSLIB Fortran wrappers completely in Fortran 2003.

         It seems that this option may be tenable as the BIND(C) spec
         allows for C-equivalent derived types.  However, it would
         require rewriting all of the wrappers, 5 kloc of C code, in
         Fortran 2003 code that must directly manipulate the internals
         of all of the WCSLIB structs.

      3) Introduce a new layer of thin wrappers, written in Fortran
         2003, that do nothing more than define the INTERFACE to the
         existing wrappers written in C and then call them.
         Essentially this extracts the changes required in option (1)
         into a new set of wrappers.

         This was the option chosen on the basis of simplicity - only
         specific wrappers, namely those with character arguments, need
         be rewrapped.  It also minimises WCSLIB's exposure to Fortran
         2003, particularly for the sake of legacy astronomical packages
         such as Miriad and AIPS.  Further, it admits the possibility of
         the optional use of this extra layer of wrappers.

    The new Fortran 2003 wrappers reside in the Fortran subdirectory in
    files by the name of *_bindc.f90.  By default they are not compiled
    and not used.  To use them, WCSLIB must be configure'd with the new
    '--with-bindc' option.  This causes the new BIND(C) wrappers to be
    compiled, and the names of the old wrappers to be changed, for
    example from wcspih_() to wcspih_c().

    LTO compile problems reported by Eli Schwartz, Gentoo maintainer.

  - Added wrappers for celenq(), disenq(), linenq(), prjenq(), spcenq(),
    tabenq(), and wcsenq().

  - In the Fortran test suite, modified twcs to test WCSENQ and also
    WCSSET in bypass mode.

* PGSBOX

  - Changes mirroring those described above for the Fortran wrappers,
    the difference being that here we have a mix of C calling Fortran
    (e.g. cpgsbox() calling PGSBOX) as well as Fortran calling C (e.g.
    PGWCSL calling pgwcsl_c()).  As with the Fortran wrappers, the new
    configure option, '--with-bindc', chooses whether to use the new
    BIND(C) PGSBOX wrappers.

* Installation

  - Added '--with-bindc' as a new configure option (or BINDC=yes from
    the environment) to signal the use of the new strictly LTO-compliant
    Fortran wrappers.  See above.

  - Modified 'configure' to report the version of gcc used.  Likewise,
    for the 'show' rule in makedefs.

* User manual

  - Quelled a nuisance compiler warning in doxextr.

  - Documentation generation moved to doxygen 1.10.0 (was 1.9.8).

(For complete change log information, see https://www.atnf.csiro.au/people/mcalabre/WCS/CHANGES .)

### PR DESCRIPTION
**Pull request by mcara**
### Description

From WCSLIB release notes:
```
* C library

  - Until now, wcsset() always operated unconditionally - the wcsprm
    struct was set or reset regardless of its current state.  Likewise
    the various *set() functions for the other structs.

    However, in some situations, particularly in threaded applications,
    it is desirable to have wcsset(), etc., check the state of the
    struct and return immediately if it has already been set.  This may
    now be accomplished by setting wcsprm::flag == 1 (instead of 0)
    before calling wcsset().  Likewise for the other structs.  This sets
    a "bypass" flag within the struct itself.

  - A new function, wcsenq(), queries the state of the wcsprm struct,
    specifically whether WCSLIB is managing its memory, whether the
    struct has been set, and whether or not it is in bypass mode, and
    whether it is self-consistent.  There are corresponding functions
    for the other structs: celenq(), disenq(), linenq(), prjenq(),
    spcenq(), and tabenq().  Please refer to the WCSLIB manual.

  - In the C test suite, modified twcs to test wcsenq() and also
    wcsset() in bypass mode.

  - Quelled nuisance compiler warnings in wcsbth(), wcspih(), wcsp2s(),
    and wcshdo().
```

Related to https://github.com/astropy/astropy/issues/16245

Fixes #16452 (pllim opened that issue before seeing this PR)

Of special interest to @astrofrog @manodeep 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by manodeep** [comment]
I am still seeing multi-threaded errors (although at a much reduced rate) even with the wcs->flag set to 1. See the actions log from v8.3 - [here](https://github.com/manodeep/wcslib-build/actions/runs/9077021010)

**Comment by manodeep** [comment]
With this v8.3, setting `wcs->flag = 1` *before* calling `wcsset()` enables this special bypass behaviour where future calls to `wcsset()` immediately return (and thereby avoid executing the code section with the thread-rac). Setting `wcs->flag=0` or any other value will execute the thread-race condition. 

On a different note, I am seeing an odd compile failure on Windows [here](https://github.com/manodeep/wcslib-build/actions/runs/9079178270/job/24947869550)

```sh
Run make WCSLIB_VERSION=8.3
gcc -Wall -Wextra -O2 -fopenmp -Iwcslib-8.3/C -Iwcslib-8.3 -fsanitize=undefined -fsanitize=bounds -fsanitize=address -fsanitize-undefined-trap-on-error -fstack-protector-all test_wcs_threads.c wcslib-8.3/C/libwcs-8.3.a -lm -fopenmp -o test_wcs_threads
In file included from wcslib-8.3/C/wcslib.h:49,
                 from test_wcs_threads.c:8:
wcslib-8.3/C/wcs.h:2301:5: error: conflicting types for 'wcsset'; have 'int(struct wcsprm *)'
 2301 | int wcsset(struct wcsprm *wcs);
      |     ^~~~~~
In file included from test_wcs_threads.c:5:
D:/a/_temp/msys64/mingw64/include/string.h:183:20: note: previous declaration of 'wcsset' with type 'wchar_t *(wchar_t *, wchar_t)' {aka 'short unsigned int *(short unsigned int *, short unsigned int)'}
  183 |   wchar_t *__cdecl wcsset(wchar_t *_Str,wchar_t _Val) __MINGW_ATTRIB_DEPRECATED_MSVC2005;
      |                    ^~~~~~
make: *** [Makefile:14: test_wcs_threads] Error 1
Error: Process completed with exit code 2.
```
Apparently, `string.h` on Windows contains a function called `wcsset`, and that is throwing the compiler off. This presumably occurred before too and has been solved within astropy.

**Comment by astrofrog** [comment]
@manodeep I have a branch trying to use the new bypass and checksum stuff and think I may have managed to get rid of the multi threading issue. Will open a PR once this is merged.

**Comment by pllim** [comment]
Do we backport wcslib bumps?

**Comment by pllim** [comment]
I am also running the optional jobs, just in case.

**Comment by astrofrog** [comment]
@pllim I'm not sure about usual process but in this case I think we should because it is going to help solve some multi-threading bugs.

**Comment by pllim** [comment]
aarch64 is taking too long and given other "exotic" archs passed, let's merge. Thanks!

**Comment by pllim** [comment]
@astrofrog , but the change log fragment is "feature" not "bugfix". Not sure if that makes sense for a backport change log.

**Comment by mcara** [comment]
Let me know whether you would like me to change the log from `feature` to `bug`. Just because something does not work well for a particular use does not necessarily mean a "bug" (to me): one cannot foresee all possible uses. Calabretta's use of `flag` does not seem to be incorrect/buggy. For this reason I thought this is not a bug.

**Comment by astrofrog** [comment]
Maybe it should be 'other'

**Comment by astrofrog** [comment]
From an astropy user's perspective this PR is neither a bugfix nor a feature

**Comment by pllim** [comment]
"other" sounds good to me. @mcara , can you please make that change? The catch is "other" cannot be in sub-dir, so you have to move it out of "wcs" (https://github.com/astropy/astropy/tree/main/docs/changes).

File: docs/changes/16451.other.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Updated bundled WCSLIB version to 8.3. This update changes the behavior of various ``*set`` functions in order to improve stability of WCSLIB in threaded applications. For a full list of changes - see
+http://www.atnf.csiro.au/people/mcalabre/WCS/CHANGES or
```
Comment (pllim) [inline docs/changes/16451.other.rst:2]: Let's remove the URL. I don't want to have to "change history" in the future when the URL breaks. Thanks!
```suggestion
```

Comment (astrofrog) [reply docs/changes/16451.other.rst:2]: Maybe let's not re-run all the extra CI just for this change though :sweat_smile:

Comment (mcara) [reply docs/changes/16451.other.rst:2]: @pllim https://github.com/astropy/astropy/blob/main/docs/wcs/references.txt is full of linksthat could change in the future. I think links are helpful and make life easier for those who read the docs and probably we will have to fix links when they become outdated.

Comment (pllim) [reply docs/changes/16451.other.rst:2]: The live links in change log topic had been debated before and we decided to not do it.

Yes, no need to rerun CI, you can do a `[ci skip]` and I can merge it.

**Comment by astrofrog** [comment]
Thanks! Looks good now.

**Pull request merged**